### ISSUE TITLE
Refresh page

### ISSUE THREAD

**Issue by CapitanLiteral** [DIGITALCRIMINAL/OnlyFans#1119]
I fill the auth.json but when I start the script it says that I should refresh the page. Does this happen to anyone else? Is this a bug or I'm doing something wrong?

**Comment by ProjektPremium** [comment]
> 
> 
> I fill the auth.json but when I start the script it says that I should refresh the page. Does this happen to anyone else? Is this a bug or I'm doing something wrong?

Just started doing it for me again also. Hold tight, it will get looked at.

**Comment by sonicshake** [comment]
> > > > The fourth part of the sign that's changing - previously 608c48da and now 608fcf2c - this looks like a hexidecimal version of an epoch timestamp in seconds. I don't have the previous vendor.js to look back at, but this is what I've observed from the new file:
> > > > 
> > > > 1. 608fcf2c converted to decimal and then analyzed as a timestamp is **2021-05-03 10:23**:40 UTC
> > > > 2. The vendor.js full file name from Chrome developer tools is vendor.js?rev=**202105031022**-6ba1f1c47b
> > > > 
> > > > It looks like the revision number and the hex code that is being used are similar but off by a minute, so maybe this is just a coincidence. Just an observation.
> > > 
> > > 
> > > I was going to say if 4: is the day and :608 is the revision then it might be dynamically generated?
> > 
> > 
> > The first number so far does seem to be incrementing, but is it by day? It seems the 3: lasted both Saturday and Sunday.
> 
> **Yeah, but on the 3rd they changed the number to 4 - maybe today they'll change it to 5.**

This just happened, lol

**Comment by trevdilley** [comment]
@DIGITALCRIMINAL, @hippothon we back at it again.

4 = 5
608fcf2c  = 6091065f
yoCCrPVmrN27vlUEQLZcW3DZH97KRVoy = Sx7FEcC7r5uKuCIzVljwS8gnZGhNprM5

need a new checksum :(

**Comment by ProjektPremium** [comment]
> 
> 
> @DIGITALCRIMINAL, @hippothon we back at it again.
> 
> 4 = 5
> 608fcf2c = 6091065f
> yoCCrPVmrN27vlUEQLZcW3DZH97KRVoy = Sx7FEcC7r5uKuCIzVljwS8gnZGhNprM5
> 
> need a new checksum :(

You called it brother.

**Comment by DIGITALCRIMINAL** [comment]
Well at least we know it's automatic and not some guy on a computer updating it manually lmao.

**Comment by idiotabroad** [comment]
Yep, it went again

I think they are detcting use of this tool somehow

**Comment by ProjektPremium** [comment]
> 
> 
> Yep, it went again
> 
> I think they are detcting use of this tool somehow

They aren't detecting that people are using it, they are probably aware of this page and have their own copy of the scraper and are testing it to see what breaks it, how it gets fixed, how quick it gets fixed etc etc.

**Comment by MrCheeseLol** [comment]
So is it borked for the time being? I've tried the updated login thingo (I'm real good with the technical lingo), it eventually lets me log in through the geckodriver browser but I then just get:

Scraping Paid Content
Scraping Subscriptions
There's nothing to scrape.
Archive Completed in 2.23 Minutes

**Comment by ProjektPremium** [comment]
> 
> 
> So is it borked for the time being? I've tried the updated login thingo (I'm real good with the technical lingo), it eventually lets me log in through the geckodriver browser but I then just get:
> 
> Scraping Paid Content
> Scraping Subscriptions
> There's nothing to scrape.
> Archive Completed in 2.23 Minutes

Try setting your max_threads to 1 and not -1

**Comment by MrCheeseLol** [comment]
Already done and no dice.

**Comment by dearcoding** [comment]
I have an idea, what if we run selenium wire just to get the correct header?

I mean for the generation of sign we use selenium, with selenium wire all is easy interceptable.

I think this can be done, but on what parameters depends sign?

**Comment by ProjektPremium** [comment]
> 
> 
> I have an idea, what if we run selenium wire just to get the correct header?
> 
> I mean for the generation of sign we use selenium, with selenium wire all is easy interceptable.
> 
> I think this can be done, but on what parameters depends sign?

Sounds like a good idea but the problem is for the people who aren't tech savvy and can't get all these things going. Just a thought you know?

**Comment by ofext** [comment]
> > I have an idea, what if we run selenium wire just to get the correct header?
> > I mean for the generation of sign we use selenium, with selenium wire all is easy interceptable.
> > I think this can be done, but on what parameters depends sign?
> 
> Sounds like a good idea but the problem is for the people who aren't tech savvy and can't get all these things going. Just a thought you know?

Forgive my ignorance, but how would selenium help in getting the header for the request? It is computed by OnlyFans site itself and changing everyday. How would we extract the computing function from their source code just by using selenium?

**Comment by dearcoding** [comment]
> > > I have an idea, what if we run selenium wire just to get the correct header?
> > > I mean for the generation of sign we use selenium, with selenium wire all is easy interceptable.
> > > I think this can be done, but on what parameters depends sign?
> > 
> > Sounds like a good idea but the problem is for the people who aren't tech savvy and can't get all these things going. Just a thought you know?
> 
> Forgive my ignorance, but how would selenium help in getting the header for the request? It is computed by OnlyFans site itself and changing everyday. How would we extract the computing function from their source code just by using selenium?

Well, the sign is generated for every request, for example it is generated also for the login page.

It is based on time (not a problem), other cache variables like auth-id and xbc, which we can edit before making the JavaScript calculate the right headers, the only problem could be if the url is used too in the calculation of the sign.

So we open our selenium sessions on only fans, we edit our cache -> we refresh page, before request is sent we get the sign value, we then abort the request.

Now we have the sign to be used in our requests.

In this way we could have always the updated version of the sign since what we are doing is just using the last JavaScript from only fans servers.

I will try to make it and will let you know where I break, but this seems the only way to have always updated headers without moving all on selenium.

**Comment by ofext** [comment]
@dearcoding the url path is used in calculating the sha1 hash, and the hash is then use to compute the thirst parameter of the sign, so it’s not a value you can get once and then use it multiple times...

**Comment by dearcoding** [comment]
> @dearcoding the url path is used in calculating the sha1 hash, and the hash is then used to compute the third parameter of the sign, so it’s not a value you can get once and then use it multiple times...

Yeah this means if we find the way to manipulate the url used in the calculation the job is done.

**Comment by ofext** [comment]
> > @dearcoding the url path is used in calculating the sha1 hash, and the hash is then used to compute the third parameter of the sign, so it’s not a value you can get once and then use it multiple times...
> 
> Yeah this means if we find the way to manipulate the url used in the calculation the job is done.

The url used is not the browser’s url, it’s a url handled internally to query the api depending on the content currently being displayed on the site and the new content being required by the user while scrolling down a page...

Not saying it can’t be done, just saying it would be kind of messy and error prone

**Comment by Macmasteri** [comment]
Got this error too. Sadly I was busy working when the script worked.

**Comment by dearcoding** [comment]
> > > @dearcoding the url path is used in calculating the sha1 hash, and the hash is then used to compute the third parameter of the sign, so it’s not a value you can get once and then use it multiple times...
> > 
> > 
> > Yeah this means if we find the way to manipulate the url used in the calculation the job is done.
> 
> The url used is not the browser’s url, it’s a url handled internally to query the api depending on the content currently being displayed on the site and the new content being required by the user while scrolling down a page...
> 
> Not saying it can’t be done, just saying it would be kind of messy and error prone

I don't know, i't just a frontend, usually this kind of controls are handled by backend.

**Comment by dukedward** [comment]
> Well at least we know it's automatically updated every 24 hours and not some guy on a computer updating it manually lmao.

so correct me if I'm wrong @DIGITALCRIMINAL, @hippothon & @trevdilley  but this appears to be what our daily challenge will be in figuring out the sign:

sign example - 5:########################################:932:6091065f

5: <--- appears to increment daily 
########################################: <--- created from combo of random daily static string/ salt, epoch timestamp, api path, userId all separated with "\n" and converted to sha1
932: <---- sha1 checksum
6091065f <--- hex conversion of vendor.js revision epoch timestamp

I wish I could be more help but I'm not very good at JS and def don't know how to reverse engineer it just doing my part to help

**Comment by ofext** [comment]
For the guy who constantly reduces the function, here it is the new one isolated, it computes the third parameter of the sign based on the SHA1 hash:

https://jsfiddle.net/85fyu9wc/

New "constants":

var str1 = "Sx7FEcC7r5uKuCIzVljwS8gnZGhNprM5";
var str2 = "6091065f";
var constNumber = 5;

**Comment by hippothon** [comment]
I have doubts that their change is automated.

Yesterday it stopped accepting the old signature around 17:00 UTC, with the js being updated around 10:22 UTC (from filename).
Today it stopped accepting the old signature around 09:00 UTC, with the js being updated around 08:30 UTC.

The discrepancy in client-side and server-side changes and the difference in times on different days tells me some guy was tasked with watching this github and trying to stay ahead of it.

Anyway the changes are

```python
    static_param = "Sx7FEcC7r5uKuCIzVljwS8gnZGhNprM5"
```

```python
    checksum = sum([sha_1_b[15], sha_1_b[37], sha_1_b[6], sha_1_b[9], sha_1_b[13], sha_1_b[34], sha_1_b[17], sha_1_b[14], sha_1_b[1], sha_1_b[37], sha_1_b[14], sha_1_b[18], sha_1_b[24], sha_1_b[28], sha_1_b[1], sha_1_b[31], 
                    sha_1_b[13], sha_1_b[14], sha_1_b[15], sha_1_b[19], sha_1_b[9], sha_1_b[29], sha_1_b[30], sha_1_b[23], 
                    sha_1_b[16], sha_1_b[13], sha_1_b[28], sha_1_b[35],
                    sha_1_b[15], sha_1_b[23], sha_1_b[28], sha_1_b[39]])-112
```

```python
    headers["sign"] = "5:{}:{:x}:6091065f".format(
        sha_1_sign, abs(checksum))
```

Or for anyone using else js:
```javascript
hash.charCodeAt(15) + 
hash.charCodeAt(37) + 
hash.charCodeAt(6) + 
hash.charCodeAt(9) + 
hash.charCodeAt(13) + 
hash.charCodeAt(34) + 
hash.charCodeAt(17) + 
hash.charCodeAt(14) + 
hash.charCodeAt(1) + 
hash.charCodeAt(37) + 
hash.charCodeAt(14) + 
hash.charCodeAt(18) + 
hash.charCodeAt(24) + 
hash.charCodeAt(28) + 
hash.charCodeAt(1) + 
hash.charCodeAt(31) + 
hash.charCodeAt(13) + 
hash.charCodeAt(14) + 
hash.charCodeAt(15) + 
hash.charCodeAt(19) + 
hash.charCodeAt(9) + 
hash.charCodeAt(29) + 
hash.charCodeAt(30) + 
hash.charCodeAt(23) + 
hash.charCodeAt(16) + 
hash.charCodeAt(13) + 
hash.charCodeAt(28) + 
hash.charCodeAt(35) + 
hash.charCodeAt(15) + 
hash.charCodeAt(23) + 
hash.charCodeAt(28) + 
hash.charCodeAt(39) +
-112
```

**Comment by DIGITALCRIMINAL** [comment]
> I have doubts that their change is automated.
> 
> Yesterday it stopped accepting the old signature around 17:00 UTC, with the js being updated around 10:22 UTC (from filename).
> Today it stopped accepting the old signature around 09:00 UTC, with the js being updated around 08:30 UTC.
> 
> The discrepancy in client-side and server-side changes and the difference in times on different days tells me some guy was tasked with watching this github and trying to stay ahead of it.
> 
> Anyway the changes are
> 
> ```python
>     static_param = "Sx7FEcC7r5uKuCIzVljwS8gnZGhNprM5"
> ```
> 
> ```python
>     checksum = sum([sha_1_b[15], sha_1_b[37], sha_1_b[6], sha_1_b[9], sha_1_b[13], sha_1_b[34], sha_1_b[17], sha_1_b[14], sha_1_b[1], sha_1_b[37], sha_1_b[14], sha_1_b[18], sha_1_b[24], sha_1_b[28], sha_1_b[1], sha_1_b[31], 
>                     sha_1_b[13], sha_1_b[14], sha_1_b[15], sha_1_b[19], sha_1_b[9], sha_1_b[29], sha_1_b[30], sha_1_b[23], 
>                     sha_1_b[16], sha_1_b[13], sha_1_b[28], sha_1_b[35],
>                     sha_1_b[15], sha_1_b[23], sha_1_b[28], sha_1_b[39]])-112
> ```
> 
> ```python
>     headers["sign"] = "5:{}:{:x}:6091065f".format(
>         sha_1_sign, abs(checksum))
> ```
> 
> Or for anyone using else js:
> 
> ```js
> hash.charCodeAt(15) + 
> hash.charCodeAt(37) + 
> hash.charCodeAt(6) + 
> hash.charCodeAt(9) + 
> hash.charCodeAt(13) + 
> hash.charCodeAt(34) + 
> hash.charCodeAt(17) + 
> hash.charCodeAt(14) + 
> hash.charCodeAt(1) + 
> hash.charCodeAt(37) + 
> hash.charCodeAt(14) + 
> hash.charCodeAt(18) + 
> hash.charCodeAt(24) + 
> hash.charCodeAt(28) + 
> hash.charCodeAt(1) + 
> hash.charCodeAt(31) + 
> hash.charCodeAt(13) + 
> hash.charCodeAt(14) + 
> hash.charCodeAt(15) + 
> hash.charCodeAt(19) + 
> hash.charCodeAt(9) + 
> hash.charCodeAt(29) + 
> hash.charCodeAt(30) + 
> hash.charCodeAt(23) + 
> hash.charCodeAt(16) + 
> hash.charCodeAt(13) + 
> hash.charCodeAt(28) + 
> hash.charCodeAt(35) + 
> hash.charCodeAt(15) + 
> hash.charCodeAt(23) + 
> hash.charCodeAt(28) + 
> hash.charCodeAt(39) +
> -112
> ```


Ahh true, lmao. Since you're here, how is the -112 calculated?

**Comment by ofext** [comment]
> I have doubts that their change is automated.
> 
> Yesterday it stopped accepting the old signature around 17:00 UTC, with the js being updated around 10:22 UTC (from filename).
> Today it stopped accepting the old signature around 09:00 UTC, with the js being updated around 08:30 UTC.
> 
> The discrepancy in client-side and server-side changes and the difference in times on different days tells me some guy was tasked with watching this github and trying to stay ahead of it.
> 
If that's the case then we must change it thousands of times until they get tired of changing it. It's not healthy to constantly change a production site's frontend and backend for reasons not related to improving user's experience or provide new  features. Unless it's autommatically generated, they must stop at some point

**Comment by trevdilley** [comment]
yeah lets just wear this dude down :)

**Comment by ProjektPremium** [comment]
I love how everybody comes together to solve the problems.

**Comment by MrCheeseLol** [comment]
You guys are fucking Wizards.

**Comment by hippothon** [comment]
> Ahh true, lmao. Since you're here, how is the -112 calculated?

The initial code after deobfuscation looks something like this:
```javascript
hash.charCodeAt(15) - 142 + 
hash.charCodeAt(37) - 124 + 
hash.charCodeAt(6) - 147 + 
hash.charCodeAt(9) - 84 + 
hash.charCodeAt(13) + 83 + 
hash.charCodeAt(34) - 101 + 
hash.charCodeAt(17) - 76 + 
hash.charCodeAt(14) + 124 + 
hash.charCodeAt(1) + 107 + 
hash.charCodeAt(37) + 151 + 
hash.charCodeAt(14) - 147 + 
hash.charCodeAt(18) - 79 + 
hash.charCodeAt(24) + 90 + 
hash.charCodeAt(28) - 59 + 
hash.charCodeAt(1) + 121 + 
hash.charCodeAt(31) - 98 + 
hash.charCodeAt(13) + 119 + 
hash.charCodeAt(14) - 77 + 
hash.charCodeAt(15) - 84 + 
hash.charCodeAt(19) - 72 + 
hash.charCodeAt(9) + 139 + 
hash.charCodeAt(29) + 121 + 
hash.charCodeAt(30) - 79 + 
hash.charCodeAt(23) + 135 + 
hash.charCodeAt(16) - 83 + 
hash.charCodeAt(13) + 69 + 
hash.charCodeAt(28) - 83 + 
hash.charCodeAt(35) + 89 + 
hash.charCodeAt(15) - 98 + 
hash.charCodeAt(23) - 76 + 
hash.charCodeAt(28) + 148 + 
hash.charCodeAt(39) + 101
```

After I posted that the first time someone pointed out it makes more sense to simplify the math so you just add all of those numbers together.

**Comment by ProjektPremium** [comment]
Can verify it's all working again, except max_threads being anything other than "1"

**Comment by benhacka** [comment]
> ```python
>     static_param = "Sx7FEcC7r5uKuCIzVljwS8gnZGhNprM5"
> ```
> 
> ```python
>     checksum = sum([sha_1_b[15], sha_1_b[37], sha_1_b[6], sha_1_b[9], sha_1_b[13], sha_1_b[34], sha_1_b[17], sha_1_b[14], sha_1_b[1], sha_1_b[37], sha_1_b[14], sha_1_b[18], sha_1_b[24], sha_1_b[28], sha_1_b[1], sha_1_b[31], 
>                     sha_1_b[13], sha_1_b[14], sha_1_b[15], sha_1_b[19], sha_1_b[9], sha_1_b[29], sha_1_b[30], sha_1_b[23], 
>                     sha_1_b[16], sha_1_b[13], sha_1_b[28], sha_1_b[35],
>                     sha_1_b[15], sha_1_b[23], sha_1_b[28], sha_1_b[39]])-112
> ```
> 
> ```python
>     headers["sign"] = "5:{}:{:x}:6091065f".format(
>         sha_1_sign, abs(checksum))
> ```
> 
> Or for anyone using else js:
> 
> ```js
> hash.charCodeAt(15) + 
> ...
> hash.charCodeAt(39) +
> -112
> ```

Awesome. But where did you get salt, and sha1 bytes order?  Is it possible to parser from JS path/text (with regular exp) with static requests?

**Comment by trevdilley** [comment]
@DIGITALCRIMINAL might as well keep this issue open. talk to you guys in a day....

**Comment by dukedward** [comment]
> @DIGITALCRIMINAL might as well keep this issue open. talk to you guys in a day....

I was just thinking the same thing... see you guys again tomorrow

**Comment by ProjektPremium** [comment]
> 
> 
> @DIGITALCRIMINAL might as well keep this issue open. talk to you guys in a day....

Take it easy, Pal.

**Comment by 8steelbeans** [comment]
> 
> 
> Can verify it's all working again, except max_threads being anything other than "1"

Hey there. Long time listener, first time caller. I'm confused. I'm still getting the same error as an hour ago.. it's working on your end?

**Comment by ProjektPremium** [comment]
> > Can verify it's all working again, except max_threads being anything other than "1"
> 
> Hey there. Long time listener, first time caller. I'm confused. I'm still getting the same error as an hour ago.. it's working on your end?

Use the latest commit.

**Comment by 8steelbeans** [comment]
> 
> 
> > > Can verify it's all working again, except max_threads being anything other than "1"
> > 
> > 
> > Hey there. Long time listener, first time caller. I'm confused. I'm still getting the same error as an hour ago.. it's working on your end?
> 
> Use the latest commit.

Ahh... of course. Thank you

**Comment by greatmate98** [comment]
Got it back working.  But its only showing 12 of the 100+ subscriptions i have.  And after each scrape, its hanging on the downloading messages part.

**Comment by wstan1** [comment]
> 
> 
> > > Can verify it's all working again, except max_threads being anything other than "1"
> > 
> > 
> > Hey there. Long time listener, first time caller. I'm confused. I'm still getting the same error as an hour ago.. it's working on your end?
> 
> Use the latest commit.

i am using latest commit but still having the refresh page issue on my end

**Comment by dilemmax** [comment]
Managed to sign in but when I try to scrape, I'm getting this

Type: Profile
0.00B [00:00, ?B/s]
Type: Stories
No Stories Found.
Type: Posts
Scrape Attempt: 1/100
Missing 100 Posts... Retrying...
Scrape Attempt: 2/100
Missing 50 Posts... Retrying...
Scrape Attempt: 3/100
Missing 50 Posts... Retrying...
Scrape Attempt: 4/100

**Comment by dilemmax** [comment]
> Can verify it's all working again, except max_threads being anything other than "1"

Did this and I was getting the above so I changed it to 1 and it's working again.

**Comment by salamihawk** [comment]
I had this issue, updated to the latest version and it works again, however:

One onlyfans model had a post mixed with images and videos. The images downloaded fine, the videos didn't.

I tried re-running the script and now it appears to be hanging on that model's videos with "0it [00:00, ?it/s]"

I've seen something about max_threads being some kind of cure-all, but where does it get set? I have it set in the onlyfans section of the config.json dict, is that right?

**Comment by dearcoding** [comment]
I'm at a good point in using selenium to get the sign value.

It's actually a good and working way as far as i tested it.

The bot will become slower because for every request i need 3/5 seconds to generate the sign, but in this way they can update their algorithm as much as they want, i will always have the right sign.

I won't run it completely in selenium because selenium is shit, memory heavy and hard to manage... 

I use selenium only for the sign calculation.

**Comment by dilemmax** [comment]
> I had this issue, updated to the latest version and it works again, however:
> 
> One onlyfans model had a post mixed with images and videos. The images downloaded fine, the videos didn't.
> 
> I tried re-running the script and now it appears to be hanging on that model's videos with "0it [00:00, ?it/s]"
> 
> I've seen something about max_threads being some kind of cure-all, but where does it get set? I have it set in the onlyfans section of the config.json dict, is that right?

@salamihawk You change it at Line 8 in config.json.

**Comment by hippothon** [comment]
> Awesome. But where did you get salt, and sha1 bytes order? Is it possible to parser from JS path/text (with regular exp) with static requests?

It's just all from removing their obfuscation.
For example this highlighted bit:
![image](https://user-images.githubusercontent.com/29172533/117031354-7e788480-acce-11eb-92e7-3c0627dded06.png)

Translates to
`e.charCodeAt(2434 % e.length) - 101`
Or since we know it's SHA1 that would be
`e.charCodeAt(2434 % 40) - 101`
Which in the snippet I posted above is then
`e.charCodeAt(34) - 101`

It's likely possible to do it statically, I have it 99% automated using regex but it runs in the browser so I can use their text replacement functions without having to copy them and mess with them myself.
The risk with regex is they change something substantially and you have to start over again.

Realistically right now it takes me about 5 min to update my code after noticing it's broken.
If it keep changing every day I might be more motivated to fully automate it but that's up to them ¯\\_(ツ)_/¯

**Comment by mediaburnwayne** [comment]
So are we just waiting on a new build?

**Comment by benhacka** [comment]
> It's just all from removing their obfuscation.
> For example this highlighted bit:
> ![image](https://user-images.githubusercontent.com/29172533/117031354-7e788480-acce-11eb-92e7-3c0627dded06.png)
> 
> Translates to
> `e.charCodeAt(2434 % e.length) - 101`
> Or since we know it's SHA1 that would be
> `e.charCodeAt(2434 % 40) - 101`
> Which in the snippet I posted above is then
> `e.charCodeAt(34) - 101`
> 

Pretty good. Thanks!

**Comment by salamihawk** [comment]
> > I had this issue, updated to the latest version and it works again, however:
> > One onlyfans model had a post mixed with images and videos. The images downloaded fine, the videos didn't.
> > I tried re-running the script and now it appears to be hanging on that model's videos with "0it [00:00, ?it/s]"
> > I've seen something about max_threads being some kind of cure-all, but where does it get set? I have it set in the onlyfans section of the config.json dict, is that right?
> 
> @salamihawk You change it at Line 8 in config.json.

Gotcha, thanks... I was still working with an old file from an old version before the auth config got split off to .profiles

Still seems to hang at the same spot though

**Comment by hippothon** [comment]
So I think it's obvious that they're watching this page since they just pushed an update that tries to interfere with you using devtools.
I'll continue to share changes but I'd advise against anyone sharing specific methods.

**Comment by dearcoding** [comment]
> So I think it's obvious that they're watching this page since they just pushed an update that tries to interfere with you using devtools.
> I'll continue to share changes but I'd advise against anyone sharing specific methods.
> 
> I don't have access to my code right now but will post the update for version 6 later.

I have a private method to compute sign using selenium, I want help community but i don't want publish it here since it would get patched quickly.

If anyone is interested please contact me on email (you get it on my profile).

**Comment by ofext** [comment]
> So I think it's obvious that they're watching this page since they just pushed an update that tries to interfere with you using devtools.
> I'll continue to share changes but I'd advise against anyone sharing specific methods.
> 
> I don't have access to my code right now but will post the update for version 6 later.

What??? Those bastards!

**Comment by WellSomething** [comment]
> > > I had this issue, updated to the latest version and it works again, however:
> > > One onlyfans model had a post mixed with images and videos. The images downloaded fine, the videos didn't.
> > > I tried re-running the script and now it appears to be hanging on that model's videos with "0it [00:00, ?it/s]"
> > > I've seen something about max_threads being some kind of cure-all, but where does it get set? I have it set in the onlyfans section of the config.json dict, is that right?
> > 
> > 
> > @salamihawk You change it at Line 8 in config.json.
> 
> Gotcha, thanks... I was still working with an old file from an old version before the auth config got split off to .profiles
> 
> Still seems to hang at the same spot though

I'm also having this problem with the 0it [00:00, ?it/s]. It only began when we got the new workarounds. I'm using v7 (which was working perfectly) and replaced the "apis" folder after the recent changes...

0it [00:00, ?it/s]Traceback (most recent call last):
  File "C:\XXXXXX\OnlyFans\OnlyFans-v7\start_ofd.py", line 43, in <module>
    apis = main_datascraper.start_datascraper(json_config, site_name_lower)
  File "C:\XXXXXX\OnlyFans\OnlyFans-v7\datascraper\main_datascraper.py", line 105, in start_datascraper
    names = main_helper.process_names(
  File "C:\XXXXXX\OnlyFans\OnlyFans-v7\helpers\main_helper.py", line 629, in process_names
    result = module.start_datascraper(
  File "C:\XXXXXX\OnlyFans\OnlyFans-v7\modules\onlyfans.py", line 156, in start_datascraper
    results = prepare_scraper(
  File "C:\XXXXXX\OnlyFans\OnlyFans-v7\modules\onlyfans.py", line 670, in prepare_scraper
    profile_scraper(authed, site_name, api_type, username,
  File "C:\XXXXXX\OnlyFans\OnlyFans-v7\modules\onlyfans.py", line 296, in profile_scraper
    r = OnlyFans.api_helper.json_request(media_link, session, stream=True,
AttributeError: module 'apis.api_helper' has no attribute 'json_request'

**Comment by Dered118** [comment]
It's working for me, scraping as we speak

**Comment by trevdilley** [comment]
@DIGITALCRIMINAL @hippothon 
Looks like we are back at it again boys

5 = 6
6091065f = 609184ae

Currently having issues finding the hash string
New checksum

**Comment by hippothon** [comment]
Here's 6

```python
    static_param = "rhtNVxJh2LD3Jul5MhHcAAnFMysnLlct"
```
```python
    checksum = sum([sha_1_b[31], sha_1_b[13], sha_1_b[8], sha_1_b[3], sha_1_b[25], sha_1_b[8], sha_1_b[33], sha_1_b[25], sha_1_b[1], sha_1_b[23], sha_1_b[37], sha_1_b[11], sha_1_b[2], sha_1_b[29], sha_1_b[9], sha_1_b[7],
					sha_1_b[29], sha_1_b[30], sha_1_b[18], sha_1_b[25], sha_1_b[18], sha_1_b[21], sha_1_b[10], sha_1_b[37],
					sha_1_b[28], sha_1_b[35], sha_1_b[31], sha_1_b[5],
					sha_1_b[13], sha_1_b[31], sha_1_b[2], sha_1_b[9]]) + 1110
					
```
```python
    headers["sign"] = "6:{}:{:x}:609184ae".format(
        sha_1_sign, abs(checksum))
```

JS:

```js
hash.charCodeAt(31) +
hash.charCodeAt(13) +
hash.charCodeAt(8) +
hash.charCodeAt(3) +
hash.charCodeAt(25) +
hash.charCodeAt(8) +
hash.charCodeAt(33) +
hash.charCodeAt(25) +
hash.charCodeAt(1) +
hash.charCodeAt(23) +
hash.charCodeAt(37) +
hash.charCodeAt(11) +
hash.charCodeAt(2) +
hash.charCodeAt(29) +
hash.charCodeAt(9) +
hash.charCodeAt(7) +
hash.charCodeAt(29) +
hash.charCodeAt(30) +
hash.charCodeAt(18) +
hash.charCodeAt(25) +
hash.charCodeAt(18) +
hash.charCodeAt(21) +
hash.charCodeAt(10) +
hash.charCodeAt(37) +
hash.charCodeAt(28) +
hash.charCodeAt(35) +
hash.charCodeAt(31) +
hash.charCodeAt(5) +
hash.charCodeAt(13) +
hash.charCodeAt(31) +
hash.charCodeAt(2) +
hash.charCodeAt(9) +
1110
```

**Comment by mstiggles** [comment]
sorry, when you guys say to use the "latest commit" what do you mean? I think I'm missing something. thanks

**Comment by badLABZ** [comment]
I'm getting the same error now, weirdly i ran it on 3 accounts and ripped them all without a problem. 
Left it 10 mins and ran again and then boom, got this error. 

The only reason I'm posting this is because i ripped them no less than 30 mins ago? this error has been up over 7+ hours.
Any chance the script is flagging to them when its run? because i don't get how it was working 30+ mins ago and i ripped 3 profiles and then it just stopped. Just a thought, I'm probably wrong. 

Also +1 to they are obviously watching this page. If i was them i would be haha

**Comment by RedMethylamine** [comment]
How did you guys debugged? There is an antidebugger running now 

> Here's 6
> 
> ```python
>     static_param = "rhtNVxJh2LD3Jul5MhHcAAnFMysnLlct"
> ```
> 
> ```python
>     checksum = sum([sha_1_b[31], sha_1_b[13], sha_1_b[8], sha_1_b[3], sha_1_b[25], sha_1_b[8], sha_1_b[33], sha_1_b[25], sha_1_b[1], sha_1_b[23], sha_1_b[37], sha_1_b[11], sha_1_b[2], sha_1_b[29], sha_1_b[9], sha_1_b[7],
> 					sha_1_b[29], sha_1_b[30], sha_1_b[18], sha_1_b[25], sha_1_b[18], sha_1_b[21], sha_1_b[10], sha_1_b[37],
> 					sha_1_b[28], sha_1_b[35], sha_1_b[31], sha_1_b[5],
> 					sha_1_b[13], sha_1_b[31], sha_1_b[2], sha_1_b[9]]) + 1110
> 					
> ```
> 
> ```python
>     headers["sign"] = "6:{}:{:x}:609184ae".format(
>         sha_1_sign, abs(checksum))
> ```
> 
> JS:
> 
> ```js
> hash.charCodeAt(31) +
> hash.charCodeAt(13) +
> hash.charCodeAt(8) +
> hash.charCodeAt(3) +
> hash.charCodeAt(25) +
> hash.charCodeAt(8) +
> hash.charCodeAt(33) +
> hash.charCodeAt(25) +
> hash.charCodeAt(1) +
> hash.charCodeAt(23) +
> hash.charCodeAt(37) +
> hash.charCodeAt(11) +
> hash.charCodeAt(2) +
> hash.charCodeAt(29) +
> hash.charCodeAt(9) +
> hash.charCodeAt(7) +
> hash.charCodeAt(29) +
> hash.charCodeAt(30) +
> hash.charCodeAt(18) +
> hash.charCodeAt(25) +
> hash.charCodeAt(18) +
> hash.charCodeAt(21) +
> hash.charCodeAt(10) +
> hash.charCodeAt(37) +
> hash.charCodeAt(28) +
> hash.charCodeAt(35) +
> hash.charCodeAt(31) +
> hash.charCodeAt(5) +
> hash.charCodeAt(13) +
> hash.charCodeAt(31) +
> hash.charCodeAt(2) +
> hash.charCodeAt(9) +
> 1110
> ```

**Comment by ofext** [comment]
> How did you guys debugged? There is an antidebugger running now
> 

I think it's best for everyone's interes if anti debugging strategies are not revealed to prevent them from attacking those methods...

**Comment by DIGITALCRIMINAL** [comment]
There's no antidebugging, if you mess up a variable you get logged out, but you're not really logged out.

**Comment by ofext** [comment]
> There's no antidebugging, if you mess up a variable you get logged out, but you're not really logged out.

In my case there's a "debugger" statement called from some infinite loop...

**Comment by mstiggles** [comment]
> > sorry, when you guys say to use the "latest commit" what do you mean? I think I'm missing something. thanks
> 
> Git clone the repo

huh?

**Comment by RedMethylamine** [comment]
I was trying to add a breakpoint on Chrome to see the output (as I've been doing with the previous versions) but now I get this:

![debugger](https://user-images.githubusercontent.com/77938230/117054664-a261b780-ad1a-11eb-8dc7-c5259cfa4ae5.png)

**Comment by mwald84** [comment]
> I was trying to add a breakpoint on Chrome to see the output (as I've been doing with the previous versions) but now I get this:
> 
> ![debugger](https://user-images.githubusercontent.com/77938230/117054664-a261b780-ad1a-11eb-8dc7-c5259cfa4ae5.png)

seems they are debugging their code. Just press to resume from debug and it works.

**Comment by Dered118** [comment]
What's the x_bc asked for in the auth or where to find that?

**Comment by Macmasteri** [comment]
Does it take longer time for you to download videos too? Been making 40 gb in over one hour and my connection is not an issue.

**Comment by NytroDev** [comment]
> What's the x_bc asked for in the auth or where to find that?

1bdaefa2c44c8690fdb32fae8e82ee0dc21f3eb4

**Comment by FelisFulgens** [comment]
> > > sorry, when you guys say to use the "latest commit" what do you mean? I think I'm missing something. thanks
> > 
> > 
> > Git clone the repo
> 
> huh?

![image](https://user-images.githubusercontent.com/83669244/117055583-f059c980-ace0-11eb-8f42-e51901e80984.png)

If you aren't using git to automatically pull the repo, go to the green code dropdown and download the latest zip file.

**Comment by idiotabroad** [comment]
The time it takes OF to bust this tool is shortening!

**Comment by Dered118** [comment]
Now getting

Auth (V1) Attempt 1/10
Session lock
.
.
.
Auth (V1) Attempt 10/10
Session lock

**Comment by NytroDev** [comment]
> Now getting
> 
> Auth (V1) Attempt 1/10
> Session lock
> .
> .
> .
> Auth (V1) Attempt 10/10
> Session lock

have you downloaded the recent commit?

**Comment by Dered118** [comment]
Yes, i'm on 7.01

**Comment by NytroDev** [comment]
> Yes, i'm on 7.01

no no not the release. do what @FelisFulgens explained the download the latest zip

**Comment by DIGITALCRIMINAL** [comment]
> > I was trying to add a breakpoint on Chrome to see the output (as I've been doing with the previous versions) but now I get this:
> > ![debugger](https://user-images.githubusercontent.com/77938230/117054664-a261b780-ad1a-11eb-8dc7-c5259cfa4ae5.png)
> 
> seems they are debugging their code. Just press to resume from debug and it works.

Oh lol, my bad.

**Comment by Dered118** [comment]
> 
> 
> > Yes, i'm on 7.01
> 
> no no not the release. do what @FelisFulgens explained the download the latest zip

Did that now, updated the cookies too, still getting Session lock

**Comment by Dered118** [comment]
When trying, i get access denied on the first login attempt and then session lock for the other 9.
After reloading the website im logged out and the sess cookie changed

**Comment by RedMethylamine** [comment]
> > > I was trying to add a breakpoint on Chrome to see the output (as I've been doing with the previous versions) but now I get this:
> > > ![debugger](https://user-images.githubusercontent.com/77938230/117054664-a261b780-ad1a-11eb-8dc7-c5259cfa4ae5.png)
> > 
> > 
> > seems they are debugging their code. Just press to resume from debug and it works.
> 
> Oh lol, my bad.

no problem. I tried the solution of @mwald84 but it stops in the same place. The partial solution I've found is to right click on the line debugger and select 'Never pause here'

Adds an extra step but seems to work

**Comment by salamihawk** [comment]
I just cloned the latest repository and my issue with frozen downloads has been resolved. Dunno what changed. Just FYI, I know it's got nothing to do with this thread :)

**Comment by RedMethylamine** [comment]
> > Ahh true, lmao. Since you're here, how is the -112 calculated?
> 
> The initial code after deobfuscation looks something like this:
> 
> ```js
> hash.charCodeAt(15) - 142 + 
> hash.charCodeAt(37) - 124 + 
> hash.charCodeAt(6) - 147 + 
> hash.charCodeAt(9) - 84 + 
> hash.charCodeAt(13) + 83 + 
> hash.charCodeAt(34) - 101 + 
> hash.charCodeAt(17) - 76 + 
> hash.charCodeAt(14) + 124 + 
> hash.charCodeAt(1) + 107 + 
> hash.charCodeAt(37) + 151 + 
> hash.charCodeAt(14) - 147 + 
> hash.charCodeAt(18) - 79 + 
> hash.charCodeAt(24) + 90 + 
> hash.charCodeAt(28) - 59 + 
> hash.charCodeAt(1) + 121 + 
> hash.charCodeAt(31) - 98 + 
> hash.charCodeAt(13) + 119 + 
> hash.charCodeAt(14) - 77 + 
> hash.charCodeAt(15) - 84 + 
> hash.charCodeAt(19) - 72 + 
> hash.charCodeAt(9) + 139 + 
> hash.charCodeAt(29) + 121 + 
> hash.charCodeAt(30) - 79 + 
> hash.charCodeAt(23) + 135 + 
> hash.charCodeAt(16) - 83 + 
> hash.charCodeAt(13) + 69 + 
> hash.charCodeAt(28) - 83 + 
> hash.charCodeAt(35) + 89 + 
> hash.charCodeAt(15) - 98 + 
> hash.charCodeAt(23) - 76 + 
> hash.charCodeAt(28) + 148 + 
> hash.charCodeAt(39) + 101
> ```
> 
> After I posted that the first time someone pointed out it makes more sense to simplify the math so you just add all of those numbers together.

btw, how do you know when to add and when to substract? I'm not able to find any criteria that makes sense

**Comment by mstiggles** [comment]
thanks felisfugens

seems to be working for me now. at least I have access and the list of models

**Comment by j-ktz** [comment]
@mstiggles How did you get past the Please Refresh This Page?

**Comment by DIGITALCRIMINAL** [comment]
Well, time for me to learn javascript. I'd hate for you to keep coming back here to post @hippothon - thanks for all you've done btw :)

**Comment by mstiggles** [comment]
even though I downloaded the v7, I downloaded master
and it worked from there

**Comment by j-ktz** [comment]
@mstiggles Thx! Worked for me too.

**Comment by bigskippercy** [comment]
> So I think it's obvious that they're watching this page since they just pushed an update that tries to interfere with you using devtools.
> I'll continue to share changes but I'd advise against anyone sharing specific methods.
> 
> I don't have access to my code right now but will post the update for version 6 later.

I concur and, interestingly enough, this repo is now trending on github (whatever that's worth).

You guys are good. I figured out a decent amount of this before finding this repo, but the method here is much cleaner. Huge thanks for that. Hopefully there's not a next time, but if there is, I'll gladly chip in.

**Comment by simonrule** [comment]
im still getting refresh page on latest version

**Comment by GhostKeys** [comment]
> So I think it's obvious that they're watching this page since they just pushed an update that tries to interfere with you using devtools.
> I'll continue to share changes but I'd advise against anyone sharing specific methods.
> 
> I don't have access to my code right now but will post the update for version 6 later.

I've sent an email to DIGITALCRIMINAL and just waiting for a reply. I'm happy to set something up or people where they can only be a part of it if you understand what's going on and are actively helping. All away from public eyes.

**Comment by canariee** [comment]
@GhostKeys, 100% seems like we are making their security's job easy.  Might even be worthwhile PGP commits or adding some more files.

**Comment by GhostKeys** [comment]
> @GhostKeys, 100% seems like we are making their security's job easy. Might even be worthwhile PGP commits or adding some more files.

No need for PGP, just a quiet area for people who understand this stuff to congregate and basically do what they have been doing, collectively not in the eye of the public.

**Comment by greatmate98** [comment]
They are really trying to break this lol.  Added a script to block using dev tools overnight.

**Comment by trevdilley** [comment]
@DIGITALCRIMINAL , @hippothon the vendor file just rotated again.

7/6092b93d now

**Comment by hippothon** [comment]
I noticed the change. They haven't disabled 6 yet but I'm working on 7.
They seem to have changed one of the text replacement functions or I'm just running on low caffeine since I'm just getting garbage back. But should still be able to make it work. They removed the anti-debugger at least.

Update soon

**Comment by dukedward** [comment]
Definitely noticed I'm still able to use 6 but hopefully we're able to get ahead of 7 before 6 is disabled.

**Comment by hippothon** [comment]
7

```python
    static_param = "cyo3uhuZp5tqYuEAhaVMuXMmPpRRsBq1"
```
```python
    checksum = sum([sha_1_b[25], sha_1_b[35], sha_1_b[26], sha_1_b[34], sha_1_b[32], sha_1_b[35], sha_1_b[32], sha_1_b[6], sha_1_b[21], sha_1_b[2], sha_1_b[15], sha_1_b[9], sha_1_b[3], sha_1_b[14], sha_1_b[26], sha_1_b[26],
                    sha_1_b[5], sha_1_b[34], sha_1_b[30], sha_1_b[34], sha_1_b[30], sha_1_b[23], sha_1_b[12], sha_1_b[20],
                    sha_1_b[20], sha_1_b[26], sha_1_b[17], sha_1_b[35],
                    sha_1_b[5], sha_1_b[16], sha_1_b[37], sha_1_b[1]]) - 956
```
```python
    headers["sign"] = "7:{}:{:x}:6092b93d".format(
        sha_1_sign, abs(checksum))
```

JS:

```js
hash.charCodeAt(25) +
hash.charCodeAt(35) +
hash.charCodeAt(26) +
hash.charCodeAt(34) +
hash.charCodeAt(32) +
hash.charCodeAt(35) +
hash.charCodeAt(32) +
hash.charCodeAt(6) +
hash.charCodeAt(21) +
hash.charCodeAt(2) +
hash.charCodeAt(15) +
hash.charCodeAt(9) +
hash.charCodeAt(3) +
hash.charCodeAt(14) +
hash.charCodeAt(26) +
hash.charCodeAt(26) +
hash.charCodeAt(5) +
hash.charCodeAt(34) +
hash.charCodeAt(30) +
hash.charCodeAt(34) +
hash.charCodeAt(30) +
hash.charCodeAt(23) +
hash.charCodeAt(12) +
hash.charCodeAt(20) +
hash.charCodeAt(20) +
hash.charCodeAt(26) +
hash.charCodeAt(17) +
hash.charCodeAt(35) +
hash.charCodeAt(5) +
hash.charCodeAt(16) +
hash.charCodeAt(37) +
hash.charCodeAt(1) +
-956
```

**Comment by ofext** [comment]
Currently both 6 and 7 signs are working and disabled anti-debug, they must be doing something else behind courtins

**Comment by hippothon** [comment]
On past updates the previous code was working anywhere from 30 minutes after pushing changes to a few hours.
Disabling it right away or even just them changing it frequently is a bad user experience since real users get the "refresh page" notice too if they have the page up or leave it open in the background.

I assume they disabled anti-debug because the endless loop could wreck performance and wastes cycles, especially on a slower browser like mobile.

**Comment by dearcoding** [comment]
They can't do much... 
 Client side they can change algorithm as much as they want and the only result would be real users getting "session locked" or other weird things, and we will always reverse engineering it.

They can make it more complicated... 
Yes, but it has to be quick to compute and light so i don't think they will.

The only thing they can do is a server side check user behavior to lock session that seems to be bot.

**Comment by ofext** [comment]
@hippothon but isn’t the endless loop only activated when opening the debug console, how would it affect regular users performance?

**Comment by hippothon** [comment]
> @hippothon but isn’t the endless loop only activated when opening the debug console, how would it affect regular users performance?

They weren't detecting if you were debugging, just running the interrupt constantly.
The function was doing something like
```
setTimeout(function() {
  debugger;
}, 100);
```

If devtools isn't open then the debugger statement does nothing but the loop is still on the event queue and still takes up CPU cycles running a check each time.

**Comment by ofext** [comment]
> They weren't detecting if you were debugging, just running the interrupt constantly.
> The function was doing something like
> 
> ```
> setTimeout(function() {
>   debugger;
> }, 100);
> ```
> 
> If devtools isn't open then the debugger statement does nothing but the loop is still on the event queue and still takes up CPU cycles running a check each time.

these people are crazy, how dare they affect millions of user just to prevent a bunch of guys (few in comparisson) from debugging.?...

**Comment by trevdilley** [comment]
@DIGITALCRIMINAL yeah still looks like 6 is active with 7.

**Comment by ThrowawayAccount202005** [comment]
Hey (sorry for not using my real account but I obviously don't want any activity on a OnlyFans scrapper have associated with my real name).

Here is an idea how to make the updates of hashing easier for the user. The scrapper would fetch the frequently changing parameters from the internet (e.g. hosted here on Gitlab).

So when the scrapper starts, it would fetch e.g. `https://raw.githubusercontent.com/DIGITALCRIMINAL/OnlyFans/master/hash.json` which would contain something like:

```
{
  "static_param": "cyo3uhuZp5tqYuEAhaVMuXMmPpRRsBq1",
  "format": "7:{}:{:x}:6092b93d",
  "chechsum_indexes": [25,35,26,...],
  "checksum_constant": -956,
}
```

It would use this information to sign the requests. What do you think?

**Comment by DIGITALCRIMINAL** [comment]
> Hey (sorry for not using my real account but I obviously don't want any activity on a OnlyFans scrapper have associated with my real name).
> 
> Here is an idea how to make the updates of hashing easier for the user. The scrapper would fetch the frequently changing parameters from the internet (e.g. hosted here on Gitlab).
> 
> So when the scrapper starts, it would fetch e.g. `https://raw.githubusercontent.com/DIGITALCRIMINAL/OnlyFans/master/hash.json` which would contain something like:
> 
> ```
> {
>   "static_param": "cyo3uhuZp5tqYuEAhaVMuXMmPpRRsBq1",
>   "format": "7:{}:{:x}:6092b93d",
>   "chechsum_indexes": [25,35,26,...],
>   "checksum_constant": -956,
> }
> ```
> 
> It would use this information to sign the requests. What do you think?

Good idea, this will stop the user from having to download the entire script again. 
OF wants to create "download fatigue" for the end user of the script whereas I'm entirely fine with updating the script whenever they change their hash.

**Comment by ofext** [comment]
> Good idea, this will stop the user from having to download the entire script again.
> OF wants to create "download fatigue" for the end user of the script whereas I'm entirely fine with updating the script whenever they change their hash.

The more they change it, the more motivated we all are to play the cat and mouse game. Their motivation is economical, we do it for fun...

**Comment by smmsnab** [comment]
> > Good idea, this will stop the user from having to download the entire script again.
> > OF wants to create "download fatigue" for the end user of the script whereas I'm entirely fine with updating the script whenever they change their hash.
> 
> The more they change it, the more motivated we all are to play the cat and mouse game. Their motivation is economical, we do it for fun...

I wouldn't mind paying some subscription fee for such sign data providing service just to take the load off my mind

**Comment by hippothon** [comment]
They pushed another update ~20 min ago but we're still on 7. I had a theory that maybe it updated with each build but I guess that disproves that and they are 100% updating manually.

The last bit did change to 6092eb4c so you might want to update to help avoid detection in case they're logging it.
(In my previous testing almost any value works there - it's the hex representation of a timestamp. Older values fail but if you use the current time or anything in the future it works. Go figure.)

**Comment by ofext** [comment]
> They pushed another update ~20 min ago but we're still on 7. I had a theory that maybe it updated with each build but I guess that disproves that and they are 100% updating manually.
> 
> The last bit did change to 6092eb4c so you might want to update to help avoid detection in case they're logging it.
> (In my previous testing almost any value works there - it's the hex representation of a timestamp. Older values fail but if you use the current time or anything in the future it works. Go figure.)

If you convert it to decimal (gives 1620241228) and then try this js code:

`new Date(1620241228*1000)`

It matches the time when you say it was changed...

**Comment by hippothon** [comment]
> If you convert it to decimal (gives 1620241228) and then try this js code:
> 
> `new Date(1620241228*1000)`
> 
> It matches the time when you say it was changed...

Yep. I assume it's auto generated based on build time, it's still not clear what purpose it actually serves though. As almost any current/future time is accepted there but they're also clearly disabling based on that since if you use one of the original values it will fail.

At least it just shows it has no relation to the algorithm at all.

**Comment by oftrash** [comment]
> > Hey (sorry for not using my real account but I obviously don't want any activity on a OnlyFans scrapper have associated with my real name).
> > Here is an idea how to make the updates of hashing easier for the user. The scrapper would fetch the frequently changing parameters from the internet (e.g. hosted here on Gitlab).
> > So when the scrapper starts, it would fetch e.g. `https://raw.githubusercontent.com/DIGITALCRIMINAL/OnlyFans/master/hash.json` which would contain something like:
> > ```
> > {
> >   "static_param": "cyo3uhuZp5tqYuEAhaVMuXMmPpRRsBq1",
> >   "format": "7:{}:{:x}:6092b93d",
> >   "chechsum_indexes": [25,35,26,...],
> >   "checksum_constant": -956,
> > }
> > ```
> > 
> > 
> > It would use this information to sign the requests. What do you think?
> 
> Good idea, this will stop the user from having to download the entire script again.
> OF wants to create "download fatigue" for the end user of the script whereas I'm entirely fine with updating the script whenever they change their hash.

I love this idea.

**Comment by DIGITALCRIMINAL** [comment]
> Hey (sorry for not using my real account but I obviously don't want any activity on a OnlyFans scrapper have associated with my real name).
> 
> Here is an idea how to make the updates of hashing easier for the user. The scrapper would fetch the frequently changing parameters from the internet (e.g. hosted here on Gitlab).
> 
> So when the scrapper starts, it would fetch e.g. `https://raw.githubusercontent.com/DIGITALCRIMINAL/OnlyFans/master/hash.json` which would contain something like:
> 
> ```
> {
>   "static_param": "cyo3uhuZp5tqYuEAhaVMuXMmPpRRsBq1",
>   "format": "7:{}:{:x}:6092b93d",
>   "chechsum_indexes": [25,35,26,...],
>   "checksum_constant": -956,
> }
> ```
> 
> It would use this information to sign the requests. What do you think?

It's been implemented

https://github.com/DIGITALCRIMINAL/OnlyFans/commit/999a374f5139c6f89d4d7fa39f218be064f05afe

API requests go through session_rules function
https://github.com/DIGITALCRIMINAL/OnlyFans/blob/999a374f5139c6f89d4d7fa39f218be064f05afe/apis/onlyfans/onlyfans.py#L51

Then session_rules sends the rules to create_signed_headers
https://github.com/DIGITALCRIMINAL/OnlyFans/blob/999a374f5139c6f89d4d7fa39f218be064f05afe/apis/onlyfans/onlyfans.py#L30

**Comment by smmsnab** [comment]
@DIGITALCRIMINAL how can I send you some ETH for doing amazing work here?

**Comment by hippothon** [comment]
> Ideally hippo should create their own repo and we just import their rules .-.

I can send PRs to help keep it updated when I'm able to but I'm personally hoping OF will give up at some point so we don't have to keep updating :/

**Comment by oftrash** [comment]
I dont think we have fully figured out what 5,6 or 7: means yet but I have ways to auto send the value in realtime to any endpoint, just let me know if you guys want that.

**Comment by DIGITALCRIMINAL** [comment]
> @DIGITALCRIMINAL how can I send you some ETH for doing amazing work here?

Thanks for asking, but I can't accept at this time. In the future I'll edit the readme and create an address list.

**Comment by hippothon** [comment]
Yeah it's just a version number to tell the server which values/algorithm to use to compare the hash.
It's a common format used when storing password hashes etc.

**Comment by MrZaghar** [comment]
> > Hey (sorry for not using my real account but I obviously don't want any activity on a OnlyFans scrapper have associated with my real name).
> > Here is an idea how to make the updates of hashing easier for the user. The scrapper would fetch the frequently changing parameters from the internet (e.g. hosted here on Gitlab).
> > So when the scrapper starts, it would fetch e.g. `https://raw.githubusercontent.com/DIGITALCRIMINAL/OnlyFans/master/hash.json` which would contain something like:
> > ```
> > {
> >   "static_param": "cyo3uhuZp5tqYuEAhaVMuXMmPpRRsBq1",
> >   "format": "7:{}:{:x}:6092b93d",
> >   "chechsum_indexes": [25,35,26,...],
> >   "checksum_constant": -956,
> > }
> > ```
> > 
> > 
> > It would use this information to sign the requests. What do you think?
> 
> Good idea, this will stop the user from having to download the entire script again.
> OF wants to create "download fatigue" for the end user of the script whereas I'm entirely fine with updating the script whenever they change their hash.

can you tell us how to get static_param from debugger ?

**Comment by oftrash** [comment]
> > > Hey (sorry for not using my real account but I obviously don't want any activity on a OnlyFans scrapper have associated with my real name).
> > > Here is an idea how to make the updates of hashing easier for the user. The scrapper would fetch the frequently changing parameters from the internet (e.g. hosted here on Gitlab).
> > > So when the scrapper starts, it would fetch e.g. `https://raw.githubusercontent.com/DIGITALCRIMINAL/OnlyFans/master/hash.json` which would contain something like:
> > > ```
> > > {
> > >   "static_param": "cyo3uhuZp5tqYuEAhaVMuXMmPpRRsBq1",
> > >   "format": "7:{}:{:x}:6092b93d",
> > >   "chechsum_indexes": [25,35,26,...],
> > >   "checksum_constant": -956,
> > > }
> > > ```
> > > 
> > > 
> > > It would use this information to sign the requests. What do you think?
> > 
> > 
> > Good idea, this will stop the user from having to download the entire script again.
> > OF wants to create "download fatigue" for the end user of the script whereas I'm entirely fine with updating the script whenever they change their hash.
> 
> can you tell us how to get static_param from debugger ?

look for `return e(t)` in vendor.js when formatted and add a break point

**Comment by MrZaghar** [comment]
:) i fixed the problem
**nodejs code** :
```javascript
const fetch=require('node-fetch');
const vm = require('vm');
const sha1 = require('js-sha1');

let signInfo=async ()=>{
	var time=Date.now();
	var req=await fetch('https://static.cdn.onlyfans.com/theme/onlyfans/spa/vendor.js?rev='+time).then(e=>e.text());
	var step=req.split('Math[')[0].split('":function').pop();
	var context={};
	vm.createContext(context);
	var code='var '+step.match(/,\w=\[".*?",.*?"];.*?\(\w,\d+\)/)[0].substr(1)+';'+step.match(/\w\["\w"\]=function\(\w\){var\s(\w)=function\(\w,\w\).*?(\w)=\{\w:.*?\}};/)[0].split(/"\]=function\(\w\)\{/)[1]+'var n=r;var e="";'+'var key=function (){return '+step.match(/return \w\(\w\)},\w\["\w"]=(.*?),\w/)[1]+'};'+'var sum=function (e){return Math.abs'+req.split('Math[')[1].split('(16)')[0].replace(/[%,]e\[(r\(.*?\))\]/g, (match, index, originalString)=>{return match.replace(index,'"length"')}).replace(/\]\[(r\(.*?\))\]/g, (match, index, originalString)=>{return match.replace(index,'"charCodeAt"')}).split(/^r\(.*?\)\]/)[1].split(/\[r\(.*?\)\]/)[0]+'["toString"](16)};';
	vm.runInContext(code, context);
	return {sum:context.sum,key:context.key};
}
let createSign=async (username,userid='0')=>{
	var sign_info=await signInfo();
	var time=Date.now();
	var path='/api2/v2/users/'+username;
	var a=[sign_info.key(),time,path,userid].join('\n');
	var sign_hex =sha1(a);
	var _chk=sign_info.sum(sign_hex);
	var sign=(new Date().getDate()+1)+':'+sign_hex+':'+_chk+':'+time.toString(16);
	return {sign:sign,time:time};
}
(async ()=>{
        // here i am going to load profile info 
	var username='vmimi'; // username
	var sign=await createSign(username);
	console.log(sign);
	fetch("https://onlyfans.com/api2/v2/users/"+username, {
		  "headers": {
		    "accept": "application/json, text/plain, */*",
		    "app-token": "33d57ade8c02dbc5a333db99ff9ae26a",
		    "sign": sign.sign,
		    "time": sign.time
		  },
	}).then(e=>e.text()).then(e=>console.log(e));
})();
```

**Comment by MrZaghar** [comment]
> :) i fixed the problem
> **nodejs code** :
> still not complete yet, but the idea arrived
> 
> ```js
> const fetch=require('node-fetch');
> const vm = require('vm');
> const sha1 = require('js-sha1');
> 
> let signInfo=async ()=>{
> 	var time=Date.now();
> 	var req=await fetch('https://static.cdn.onlyfans.com/theme/onlyfans/spa/vendor.js?rev='+time).then(e=>e.text());
> 	var step=req.split('Math[')[0].split('":function').pop();
> 	var context={};
> 	vm.createContext(context);
> 	var code='var '+step.match(/,\w=\[".*?",.*?"];.*?\(\w,\d+\)/)[0].substr(1)+';'+step.match(/\w\["\w"\]=function\(\w\){var\s(\w)=function\(\w,\w\).*?(\w)=\{\w:.*?\}};/)[0].split(/"\]=function\(\w\)\{/)[1]+'var n=r;var e="";'+'var key=function (){return '+step.match(/return \w\(\w\)},\w\["\w"]=(.*?),\w/)[1]+'};'+'var sum=function (e){return Math.abs'+req.split('Math[')[1].split('(16)')[0].replace(/[%,]e\[(r\(.*?\))\]/g, (match, index, originalString)=>{return match.replace(index,'"length"')}).replace(/\]\[(r\(.*?\))\]/g, (match, index, originalString)=>{return match.replace(index,'"charCodeAt"')}).split(/^r\(.*?\)\]/)[1].split(/\[r\(.*?\)\]/)[0]+'["toString"](16)};';
> 	vm.runInContext(code, context);
> 	return {sum:context.sum,key:context.key};
> }
> let createSign=async (username,userid='0')=>{
> 	var sign_info=await signInfo();
> 	var time=Date.now();
> 	var path='/api2/v2/users/'+username;
> 	var a=[sign_info.key(),time,path,userid].join('\n');
> 	var sign_hex =sha1(a);
> 	var _chk=sign_info.sum(sign_hex);
> 	var sign=(new Date().getDate()+1)+':'+sign_hex+':'+_chk+':'+time.toString(16);
> 	return {sign:sign,time:time};
> }
> (async ()=>{
>         // here i am going to load profile info 
> 	var username='vmimi'; // username
> 	var sign=await createSign(username);
> 	console.log(sign);
> 	fetch("https://onlyfans.com/api2/v2/users/"+username, {
> 		  "headers": {
> 		    "accept": "application/json, text/plain, */*",
> 		    "app-token": "33d57ade8c02dbc5a333db99ff9ae26a",
> 		    "sign": sign.sign,
> 		    "time": sign.time
> 		  },
> 	}).then(e=>e.text()).then(e=>console.log(e));
> })();
> ```

i can change code to python but i am not fan with python enjoy

**Comment by DIGITALCRIMINAL** [comment]
> ```js
> const fetch=require('node-fetch');
> const vm = require('vm');
> const sha1 = require('js-sha1');
> ```

Actually works, good job :)

**Comment by MrZaghar** [comment]
> > i can change code to python but i am not fan with python enjoy
> 
> Actually works, good job :)

still need some changes
anyway i am going to create API it will generate the code automatic
tomorrow i am going to create it

**Comment by MrZaghar** [comment]
hello guys i created that api
userid is optional as default = 0
path is required
`https://fulldp.co/api/ofans?path=/api2/v2/users/vmimi`
if you logged with your account send this request
`https://fulldp.co/api/ofans?path=/api2/v2/users/vmimi&userid=123432`

**Comment by ixover** [comment]
hello guys i created that api
userid is optional as default = 0
path is required
`https://fulldp.co/api/ofans?path=/api2/v2/users/vmimi`
if you logged with your account send this request
`https://fulldp.co/api/ofans?path=/api2/v2/users/vmimi&userid=123432`

**Comment by oftrash** [comment]
@DIGITALCRIMINAL , @hippothon  looks like 8 is out 

looking for an update to https://raw.githubusercontent.com/DATAHOARDERS/dynamic-rules/main/onlyfans.json

**Comment by hippothon** [comment]
> @DIGITALCRIMINAL , @hippothon looks like 8 is out
> 
> looking for an update to https://raw.githubusercontent.com/DATAHOARDERS/dynamic-rules/main/onlyfans.json

Already submitted a PR.
7 still works for now.

**Comment by oftrash** [comment]
@hippothon  yeah i saw your pull request right after i commented :)

**Comment by RedMethylamine** [comment]
Man, that was fast

> > @DIGITALCRIMINAL , @hippothon looks like 8 is out
> > looking for an update to https://raw.githubusercontent.com/DATAHOARDERS/dynamic-rules/main/onlyfans.json
> 
> Already submitted a PR.
> 7 still works for now.

**Comment by whassupfr** [comment]
Someone know how works the sign hash if we're not loggin ? How can we generate the hash without the "auth_id" ?

**Comment by bigskippercy** [comment]
> Someone know how works the sign hash if we're not loggin ? How can we generate the hash without the "auth_id" ?

If you're talking about the string that's after the path, I believe it's 0 if you're not logged in

**Comment by whassupfr** [comment]
> > Someone know how works the sign hash if we're not loggin ? How can we generate the hash without the "auth_id" ?
> 
> If you're talking about the string that's after the path, I believe it's 0 if you're not logged in

Thanks It works ! But my request still receive a 400, maybe I've forget a header param ? This is my PHP code to generate the sign :

`$time = (string)round(microtime(true) * 1000);
                $path = '/api2/v2/users/onlyfans';
                $authid = '0';

                $message = utf8_encode("$staticParam\n$time\n$path\n$authid");
                $hash = sha1($message);                

                $sum = 0;
                foreach ($checksumIndexes as $checksumIndexe) {
                    $char = substr($hash, $checksumIndexe, 1);
                    $sum += ord($char);
                }
                $sum += $checksumConstant;
                $hexa = dechex($sum);

                $sign = "8:{$hash}:{$hexa}:$p3";
                
                $response = $client->request('GET', "https://onlyfans.com$path?app-token=33d57ade8c02dbc5a333db99ff9ae26a", [
                    'headers' => [
                        'accept' => 'application/json, text/plain, */*',
                        'sign' => $sign,
                        'time' => $time,
                        'x-bc' => '222dd751ad605117a5d7cb2dfbcf4360accffb39',
                        'user-agent' => 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36'
                    ],
                ]);`

**Comment by ofext** [comment]
They changed it again, yesterday they were in version 8 and now they're in version 10

**Comment by DIGITALCRIMINAL** [comment]
> They changed it again, yesterday they were in version 8 and now they're in version 10

Fixed

**Comment by oftrash** [comment]
@DIGITALCRIMINAL now its on 13

**Comment by smmsnab** [comment]
Any news on 14? Seems they have changed js generator code a bit

**Comment by DIGITALCRIMINAL** [comment]
> Any news on 14? Seems they have changed js generator code a bit

Does it not work
